torchao/quantization/README.md

```diff
@@ -164,6 +164,125 @@ model = torch.compile(model, mode='max-autotune')
 model(input)
 ```
 
+## Affine Quantization
+Affine quantization refers to the type of quantization that maps from floating point numbers to quantized numbers (typically integer) with an affine transformation, i.e.: `quantized_val = float_val / scale + zero_point` where `scale` and `zero_point` are quantization parameters for some granularity and based on some data.
+
+### Quantization Primitives
+We used to have different quantize and dequantize operators for quantization with different granularities. But in the end these can all be expressed with a `block_size` argument with different settings, so we unified existing quant primitives to `choose_qparams_affine`, `quantize_affine` and `dequantize_affine` that can represent symmetric/asymmetric per tensor/channel/token/channel_group quantization, this can be used to implement the unified quantized tensor subclass.
+
+### Quantized Tensor Subclass
+We also have a unified quantized tensor subclass that implements how to get a quantized tensor from floating point tensor and what does it mean to call linear ops on an instance of the tensor, e.g. `F.linear` and `aten.addmm`, with this we could dispatch to different operators (e.g. `int4mm` op) based on device (cpu, cuda) and quantization settings (`int4`, `int8`) and also packing formats (e.g. format optimized for cpu int4 mm kernel)
+
+### Quantization Flow
+What we need to do afterwards is roughly the following
+
+```
+for n, m in model.named_modules():
+    # or use some filter_fn
+    if isinstance(m, torch.nn.Linear):
+        # optional filtering for module name, shape etc.
+        # quantization activation (needed by dynamic quantization)
+        # m.weight = nn.Parameter(to_laq(m.weight, device=..., layout=..., ...))
+        m.weight = nn.Parameter(to_aq(m.weight, device=..., layout=..., ...))
```

Comment:
Can we make this a real snippet people can copy paste?

Reply:
yeah sure